### PR TITLE
Bug 2076619: Modified git import flow module to handle create button enable-disable issue

### PR DESCRIPTION
**Pull request by lokanandaprabhu**
**Fixes**: 
https://issues.redhat.com/browse/OCPBUGSM-43404

**Analysis / Root cause**: 
1. Build starategy was not update when builder image is selected by default, so the create button was not getting enabled.
2. No separate message when devfile is not detected for Git type - other
3. On change of import strategy error message was not cleared.

**Solution Description**: 
1. Updated Build starategy to 'Source'.
2. Disabled Devfile import strategy for Git type - other
3. On change of import strategy calling a method which will handle the error showing part.

**Screen shots / Gifs for design review**: 

https://user-images.githubusercontent.com/102503482/191943839-30293edb-8866-4867-80ae-3e1b358689fa.mov

**Unit test coverage report**: 
NA

**Test setup:**
CASE A
1. Switch to developer perspective, add page > import from Git
2. Enter URL https://mysupersecretgit.example.com/org/repo
3. Select Builder image (node.js for example), enter a name. Now all required fields are filled.

CASE B
1. Switch to developer perspective, add page > import from Git
2. Enter URL https://mysupersecretgit.example.com/org/repo
3. Select Devfile, enter "Devfile", a name. Now all required fields are filled.

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/dev-console/src/components/import/devfile/DevfileStrategySection.tsx [hunk 1]
```diff
@@ -77,12 +77,15 @@ const DevfileStrategySection: React.FC = () => {
       return t('devconsole~Validated');
     }
     if (validated === ValidatedOptions.error) {
+      if (type === GitProvider.UNSURE) {
```
Comment (abhinandan13jan) [inline frontend/packages/dev-console/src/components/import/devfile/DevfileStrategySection.tsx:80]: How are we certain that UNSURE maps to Devfile always?

File: frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        <form onSubmit={formikProps.handleSubmit}>
+          {children}
+          <input type="submit" value="Submit" />
+        </form>
+      )}
+    </Formik>
+  </Provider>
+);
+
+afterEach(() => cleanup());
```
Comment (abhinandan13jan) [inline frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx:28]: put this inside describe

File: frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+const devStrategyInput = {
```
Comment (abhinandan13jan) [inline frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx:30]: nit: You can make this reusable by putting this in external file like frontend/packages/dev-console/integration-tests/testData/git-import-flow.data.ts

**Comment by abhinandan13jan** [comment]
tested locally, works as expected... but I see that revalidation isn't working properly. Not sure whether it is related. cc: @jerolimov 

For devfile flow
revalidation


For github flow
revalidation2

File: frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx (hunk 2 above)
Comment (lokanandaprabhu) [reply frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx]: Made the changes

File: frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx (hunk 3 above)
Comment (lokanandaprabhu) [reply frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx]: Made the changes

**Comment by lokanandaprabhu** [comment]
> tested again!! the issue around revalidation still exists

I also tested this. My observation is, it is like this in 4.11 (Master) and 4.10 versions also. If URL gets validated successfully (consider for github) it is revalidating for onChange event.

I will wait for @jerolimov response on this.

cc: @jerolimov

File: frontend/packages/dev-console/src/components/import/devfile/DevfileStrategySection.tsx [hunk 4]
```diff
@@ -106,9 +109,11 @@ const DevfileStrategySection: React.FC = () => {
   }, [recommendedStrategy, setFieldValue, strategies]);
 
   React.useEffect(() => {
-    importType === 'devfile' && devfile.devfilePath && handleDevfileChange();
+    (importType === 'devfile' || selectedStrategy.type === ImportStrategy.DEVFILE) &&
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/import/devfile/DevfileStrategySection.tsx:112]: We have an enum for the importType:

```suggestion
    (importType === ImportTypes.devfile || selectedStrategy.type === ImportStrategy.DEVFILE) &&
```

File: frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        <form onSubmit={formikProps.handleSubmit}>
+          {children}
+          <input type="submit" value="Submit" />
+        </form>
+      )}
+    </Formik>
+  </Provider>
+);
+
+describe('DevFileImportStrategy', () => {
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx:28]: Doesn't this test file primary test `DevfileStrategySection` ? In this case the file should also renamed and moved into `../import/devfile/__tests__` folder and should be called `DevfileStrategySection.spec.tsx` or?

```suggestion
describe('DevfileStrategySection', () => {
```

File: frontend/packages/dev-console/integration-tests/testData/git-import-flow.data.ts [hunk 6]
```diff
@@ -13,3 +16,57 @@ export const testData = {
   scalingLimit: 2,
   label: 'app=devlelopment',
 };
+
+export const devStrategyInput = {
```
Comment (jerolimov) [inline frontend/packages/dev-console/integration-tests/testData/git-import-flow.data.ts:20]: Should be called devfile

```suggestion
export const devfileStrategyInput = {
```

File: frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,89 @@
+import * as React from 'react';
+import { render, waitFor, screen, configure, cleanup, fireEvent } from '@testing-library/react';
+import { Formik, FormikConfig } from 'formik';
+import { act } from 'react-dom/test-utils';
+import { Provider } from 'react-redux';
+import { GitProvider } from '@console/git-service/src';
+import store from '@console/internal/redux';
+import { devStrategyInput } from '../../../../../integration-tests/testData/git-import-flow.data';
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/import/git/__tests__/GitSection.spec.tsx:8]: When just used in this test case, you can also inline the mock data or create a DevfileStrategySection.data.tsx parallel to this file. These cross imports from a unit test to our e2e make it difficult to maintain this later.

**Review by jerolimov — CHANGES_REQUESTED** [review]

File: frontend/packages/dev-console/locales/en/devconsole.json [hunk 8]
```diff
@@ -374,6 +374,7 @@
   "Sample repository:": "Sample repository:",
   "The Devfile in your Git repository is invalid.": "The Devfile in your Git repository is invalid.",
   "Validated": "Validated",
+  "Could not get Devfile for an unknown Git type": "Could not get Devfile for an unknown Git type",
```
Comment (invincibleJai) [inline frontend/packages/dev-console/locales/en/devconsole.json:377]: do we need to check on the wording with UXD/PM?

cc @jerolimov

Comment (jerolimov) [reply frontend/packages/dev-console/locales/en/devconsole.json:377]: Sure! Please check with @serenamarie125 or @beaumorley

**Comment by jerolimov** [comment]
Hey @lokanandaprabhu , do get this please address the following issues:

1. Rebase this PR
2. Address Jais two review comments (add a type for devfileStrategyInput and use data-test instead of data-test-id) https://github.com/openshift/console/pull/11493#pullrequestreview-1028203535
3. Check with Serena the wording of "Could not get Devfile for an unknown Git type"

When the user selects a strategy and changes the URL (the comment from Abhi) the Import strategy always jumps to Builder Image is IMO not related to this PR. This happens also for GitHub URLs. We could improve this case, but we can address this edge case another time.

**Review by jerolimov — CHANGES_REQUESTED** [review]
@lokanandaprabhu The recommendation from @serenamarie125 was that we should "disable" the "Devfile" button if the git provider (git type) is other (and we don't support the Devfile).

It's correct that we don't need the user input for a Devfile if we know that this variant is not supported.

We can disable the button and show a tooltip that explains why the button is disabled. Maybe you can share first a draft how this would look like before you update all the tests and cleanup your code.

I'll add hold on this PR so that it's clear that this needs more work before someone reviews this again.

/hold

**Comment by lokanandaprabhu** [comment]
Hi @jerolimov ,

I attached the screenshot, if the git type is Other, then Devfile has to be disabled with a tooltip. Just for taking screenshot I added some text there, I can change that. This design is fine or any feedback on this? If this is fine, I will start to work on this.

Screenshot 2022-08-08 at 3 55 38 PM

**Comment by lokanandaprabhu** [comment]
Hi @jerolimov ,

Can you take a look at the comment - https://github.com/openshift/console/pull/11493#issuecomment-1208005946 and provide your feedback. 

Thanks.

**Comment by lokanandaprabhu** [comment]
Hi @jerolimov ,

Requesting your review on this.

File: frontend/packages/dev-console/integration-tests/support/step-definitions/addFlow/create-from-git.ts [hunk 9]
```diff
@@ -256,3 +256,19 @@ Then('user is able to see {string} value is selected in TLS termination', (value
 Then('user is able to see {string} value is selected in Insecure traffic', (value: string) => {
   cy.get(gitPO.advancedOptions.routing.insecureTraffic).should('have.text', value);
 });
+
+When('devfile import strategy is disabled', () => {
```
Comment (jerolimov) [inline frontend/packages/dev-console/integration-tests/support/step-definitions/addFlow/create-from-git.ts:260]: Nit: This is what you expect:

```suggestion
Then('devfile import strategy is disabled', () => {
```

**Comment by jerolimov** [comment]
@lokanandaprabhu Sorry for the delay :scream: I've tested this PR now and it looks good to me. I added one comment for the e2e test above. Can you rebase this PR and then I expect that we can merge it.

**Review by jerolimov — CHANGES_REQUESTED** [review]
@lokanandaprabhu Sorry for the delay :scream: I've tested this PR now and it looks good to me. I added one comment for the e2e test above. Can you rebase this PR and then I expect that we can merge it.

**Comment by lokanandaprabhu** [comment]
> @lokanandaprabhu Sorry for the delay 😱 I've tested this PR now and it looks good to me. I added one comment for the e2e test above. Can you rebase this PR and then I expect that we can merge it.

Hi @jerolimov ,

Updated the PR and rebased it. Thanks.

**Comment by jerolimov** [comment]
@lokanandaprabhu the frontend job fails because this linter issue:

```
/go/src/github.com/openshift/console/frontend/packages/dev-console/src/components/import/ImportStrategySelector.tsx
  8:1  error  `@console/git-service/src` import should occur before import of `@console/internal/actions/ui`  import/order
```

File: frontend/packages/dev-console/integration-tests/tsconfig.json [hunk 10]
```diff
@@ -5,7 +5,8 @@
     "strict": true,
     "esModuleInterop": true,
     "types": ["cypress"],
-    "lib": ["es6", "dom", "es2017"]
+    "lib": ["es6", "dom", "es2017"],
+    "jsx": "react"
```
Comment (jerolimov) [inline frontend/packages/dev-console/integration-tests/tsconfig.json:9]: Is this change really required for this PR?

File: frontend/packages/dev-console/src/components/import/git/GitSection.tsx [hunk 11]
```diff
@@ -315,6 +315,7 @@ const GitSection: React.FC<GitSectionProps> = ({
           recommendedStrategy: null,
           showEditImportStrategy: true,
         });
+        setFieldValue('build.strategy', BuildStrategyType.Source);
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/import/git/GitSection.tsx:318]: In which case is this needed? The form jumps already on an URL change to the Builder Image (Source) tab, or? But I'm okay with it as I don't see an issue here.

Generally, the form has issues to keep the state that the user has manually touched.

**Review by jerolimov — CHANGES_REQUESTED** [review]
Please check `yarn lint`. Also added two small questions:

File: frontend/packages/dev-console/src/components/import/git/GitSection.tsx (hunk 11 above)
Comment (lokanandaprabhu) [reply frontend/packages/dev-console/src/components/import/git/GitSection.tsx:318]: The above change is for below case

1. Switch to developer perspective, add page > import from Git
2. Enter URL https://mysupersecretgit.example.com/org/repo
3. Select Builder image (node.js for example), enter a name. Now all required fields are filled.

Actual results:
The user can not import the Deployment

Before :


https://user-images.githubusercontent.com/102503482/215422237-2c46437e-62bc-4827-b709-47d57691cb1d.mov

After:

https://user-images.githubusercontent.com/102503482/215422281-1e53b496-fc50-4d31-bf20-e29e751d4b09.mov

File: frontend/packages/dev-console/integration-tests/tsconfig.json (hunk 10 above)
Comment (lokanandaprabhu) [reply frontend/packages/dev-console/integration-tests/tsconfig.json]: Before I have added normal tests not the e2e tests. That time I have added this. Now I removed.

File: frontend/packages/dev-console/src/components/import/git/GitSection.tsx (hunk 11 above)
Comment (jerolimov) [reply frontend/packages/dev-console/src/components/import/git/GitSection.tsx:318]: Works fine, thank :+1:

**Review by jerolimov — APPROVED** [review]
Works fine :+1: 

/lgtm
/approve

**Comment by lokanandaprabhu** [comment]
manual cherry-pick PR #12775
